File: frontend/app/types/models/stock_movement.ts
import type { Status } from "../enums/status_enum";
import type { Warehouse } from "./warehouse";
import type { Product } from "./product";

export interface StockMovement {
  id: string;
  movement_no: string;
  type: string;
  origin_warehouse_id?: string;
  origin_warehouse?: Warehouse;
  destination_warehouse_id?: string;
  destination_warehouse?: Warehouse;
  reference_no: string;
  status: Status;
  transaction_date: string;
  note: string;
  items: StockMovementItem[];
}

export interface StockMovementItem {
  id: string;
  product_id: string;
  product?: Product;
  quantity: number;
  note: string;
}

export interface StockMovementDTO {
  id?: string;
  type: string;
  origin_warehouse_id?: string;
  destination_warehouse_id?: string;
  reference_no?: string;
  status: Status;
  transaction_date?: string;
  note?: string;
  items: StockMovementItemDTO[];
}

export interface StockMovementItemDTO {
  id?: string;
  product_id: string;
  quantity: number;
  note?: string;
}

File: backend/internal/modules/stock_movement/domain/stock_movement.go
package domain

import (
	"putra4648/erp/internal/modules/shared/enums"
	domainWarehouse "putra4648/erp/internal/modules/warehouse/domain"
	"time"

	"github.com/google/uuid"
)

type StockMovement struct {
	ID                     uuid.UUID                  `gorm:"type:uuid;primarykey;default:uuid_generate_v4()"`
	MovementNo             string                     `gorm:"type:varchar(255);unique"`
	Type                   enums.TransferType         `gorm:"type:varchar(255)"` // IN, OUT, TRANSFER
	OriginWarehouseID      *uuid.UUID                 `gorm:"type:uuid;column:origin_warehouse_id"`
	OriginWarehouse        *domainWarehouse.Warehouse `gorm:"foreignKey:OriginWarehouseID"`
	DestinationWarehouseID *uuid.UUID                 `gorm:"type:uuid;column:destination_warehouse_id"`
	DestinationWarehouse   *domainWarehouse.Warehouse `gorm:"foreignKey:DestinationWarehouseID"`
	ReferenceNo            string                     `gorm:"type:varchar(255)"`
	Status                 enums.Status               `gorm:"type:varchar(255)"`
	TransactionDate        time.Time                  `gorm:"type:timestamp"`
	Note                   string                     `gorm:"type:text"`
	Items                  []*StockMovementItem       `gorm:"foreignKey:StockMovementID"`
}
